[Bug report] IcebergTableHookDispatcher fails with NoSuchTableException when credential vending (stageCreate=true) is enabled

### Version

main branch

### Describe what's wrong

Tested this from Trino, with `iceberg.rest-catalog.vended-credentials-enabled=true`


When credential vending is enabled ,Trino sends CREATE TABLE requests with stageCreate=true. In the Iceberg REST protocol, this means the table metadata is built in memory and returned to the client along with vended credentials — the table is not committed to the catalog at this point.

However, IcebergTableHookDispatcher.createTable() unconditionally calls importTable(), which attempts to load the table from the Gravitino catalog via TableDispatcher.loadTable(). Since the table was only staged (not committed), this fails with NoSuchTableException.

The table should only be imported into Gravitino's entity store when it is actually committed — which happens later via updateTable().


### Error message and/or stacktrace



2026-04-13 19:45:47 INFO  [iceberg-rest-36] IcebergNamespaceOperations:173 - Check Iceberg namespace exists, catalog: RokuIcebergCatalog, namespace: gravitino_test_schema77
2026-04-13 19:45:47 INFO  [iceberg-rest-38] IcebergTableOperations:160 - Create Iceberg table, catalog: RokuIcebergCatalog, namespace: gravitino_test_schema77, create table request: CreateTableRequest{name=sample_events, location=gs://roku-dea-dev/sand-box/gravitino/sandbox/gravitino_test_schema/sample_events, properties={write.format.default=PARQUET, format-version=2, write.parquet.compression-codec=}, schema=table {
  1: event_id: optional long
  2: user_id: optional long
  3: event_type: optional string
  4: event_value: optional double
  5: created_at: optional timestamp
  6: date_key: optional date
}, partitionSpec=org.apache.iceberg.UnboundPartitionSpec@6b47e212, writeOrder=org.apache.iceberg.UnboundSortOrder@13691dc, stageCreate=true}, accessDelegation: vended-credentials, isCredentialVending: true
2026-04-13 19:45:47 INFO  [iceberg-rest-38] IcebergCatalogWrapper:200 - Generate credential: gcs-token for Iceberg table: gravitino_test_schema77.sample_events
2026-04-13 19:45:47 WARN  [iceberg-rest-38] IcebergExceptionMapper:84 - Iceberg REST server unexpected exception:
org.apache.gravitino.exceptions.NoSuchTableException: Iceberg table does not exist: sample_events
  at org.apache.gravitino.catalog.lakehouse.iceberg.IcebergCatalogOperations.loadTable(IcebergCatalogOperations.java:393) ~[?:?]
  at org.apache.gravitino.catalog.TableOperationDispatcher.lambda$internalLoadTable$20(TableOperationDispatcher.java:502) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.CatalogManager$CatalogWrapper.lambda$doWithTableOps$1(CatalogManager.java:153) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.utils.IsolatedClassLoader.withClassLoader(IsolatedClassLoader.java:86) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.CatalogManager$CatalogWrapper.doWithTableOps(CatalogManager.java:148) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.TableOperationDispatcher.lambda$internalLoadTable$21(TableOperationDispatcher.java:502) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.OperationDispatcher.doWithCatalog(OperationDispatcher.java:97) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.TableOperationDispatcher.internalLoadTable(TableOperationDispatcher.java:500) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.TableOperationDispatcher.lambda$loadTable$3(TableOperationDispatcher.java:117) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.lock.TreeLockUtils.doWithTreeLock(TreeLockUtils.java:49) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.TableOperationDispatcher.loadTable(TableOperationDispatcher.java:117) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.hook.TableHookDispatcher.loadTable(TableHookDispatcher.java:63) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.catalog.TableNormalizeDispatcher.loadTable(TableNormalizeDispatcher.java:63) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.listener.TableEventDispatcher.loadTable(TableEventDispatcher.java:99) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.dispatcher.IcebergTableHookDispatcher.importTable(IcebergTableHookDispatcher.java:185) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.dispatcher.IcebergTableHookDispatcher.createTable(IcebergTableHookDispatcher.java:60) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.dispatcher.IcebergTableEventDispatcher.createTable(IcebergTableEventDispatcher.java:101) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.rest.IcebergTableOperations.lambda$createTable$1(IcebergTableOperations.java:175) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at java.base/java.security.AccessController.doPrivileged(Unknown Source) ~[?:?]
  at java.base/javax.security.auth.Subject.doAs(Unknown Source) ~[?:?]
  at org.apache.gravitino.utils.PrincipalUtils.doAs(PrincipalUtils.java:44) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.server.web.Utils.doAs(Utils.java:202) ~[gravitino-server-common-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.rest.IcebergTableOperations.createTable(IcebergTableOperations.java:169) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at org.apache.gravitino.iceberg.service.rest.IcebergTableOperations_$$_jvst446_1._d2createTable(IcebergTableOperations_$$_jvst446_1.java) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
  at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
  at org.glassfish.hk2.utilities.reflection.ReflectionHelper.invoke(ReflectionHelper.java:1268) ~[hk2-utils-2.6.1.jar:?]
  at org.jvnet.hk2.internal.MethodInterceptorHandler$MethodInvocationImpl.proceed(MethodInterceptorHandler.java:164) ~[hk2-locator-2.6.1.jar:?]
  at org.apache.gravitino.server.web.filter.BaseMetadataAuthorizationMethodInterceptor.invoke(BaseMetadataAuthorizationMethodInterceptor.java:203) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at org.jvnet.hk2.internal.MethodInterceptorHandler.invoke(MethodInterceptorHandler.java:97) ~[hk2-locator-2.6.1.jar:?]
  at org.apache.gravitino.iceberg.service.rest.IcebergTableOperations_$$_jvst446_1.createTable(IcebergTableOperations_$$_jvst446_1.java) ~[gravitino-iceberg-rest-server-1.2.0.jar:?]
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
  at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
  at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:146) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:189) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:93) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.internal.Errors.process(Errors.java:292) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.internal.Errors.process(Errors.java:274) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.internal.Errors.process(Errors.java:244) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) ~[jersey-common-2.41.jar:?]
  at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:235) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684) ~[jersey-server-2.41.jar:?]
  at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394) ~[jersey-container-servlet-core-2.41.jar:?]
  at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346) ~[jersey-container-servlet-core-2.41.jar:?]
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358) ~[jersey-container-servlet-core-2.41.jar:?]
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311) ~[jersey-container-servlet-core-2.41.jar:?]
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205) ~[jersey-container-servlet-core-2.41.jar:?]
  at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.apache.gravitino.server.authentication.AuthenticationFilter.lambda$doFilter$0(AuthenticationFilter.java:89) ~[gravitino-server-common-1.2.0.jar:?]
  at java.base/java.security.AccessController.doPrivileged(Unknown Source) ~[?:?]
  at java.base/javax.security.auth.Subject.doAs(Unknown Source) ~[?:?]
  at org.apache.gravitino.utils.PrincipalUtils.doAs(PrincipalUtils.java:44) ~[gravitino-core-1.2.0.jar:?]
  at org.apache.gravitino.server.authentication.AuthenticationFilter.doFilter(AuthenticationFilter.java:86) ~[gravitino-server-common-1.2.0.jar:?]
  at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505) ~[jetty-servlet-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.Server.handle(Server.java:516) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487) ~[jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732) [jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479) [jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277) [jetty-server-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [jetty-io-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [jetty-io-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [jetty-util-9.4.51.v20230217.jar:9.4.51.v20230217]
  at java.base/java.lang.Thread.run(Unknown Source) [?:?]
Caused by: org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: gravitino_test_schema77.sample_events
  at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:55) ~[iceberg-core-1.10.1.jar:?]
  at org.apache.iceberg.rest.CatalogHandlers.loadTable(CatalogHandlers.java:329) ~[iceberg-core-1.10.1.jar:?]
  at org.apache.gravitino.iceberg.common.ops.IcebergCatalogWrapper.loadTable(IcebergCatalogWrapper.java:207) ~[gravitino-iceberg-common-1.2.0.jar:?]
  at org.apache.gravitino.catalog.lakehouse.iceberg.IcebergCatalogOperations.loadTable(IcebergCatalogOperations.java:385) ~[?:?]
  ... 93 more





### How to reproduce


Steps to reproduce:
```
CREATE SCHEMA iceberg.gravitino_test_schema77
WITH (location = '<prefix>/sand-box/gravitino/sandbox/gravitino_test_schema77/')


CREATE TABLE iceberg.gravitino_test_schema77.sample_events (
    event_id    BIGINT,
    user_id     BIGINT,
    event_type  VARCHAR,
    event_value DOUBLE,
    created_at  TIMESTAMP(6),
    date_key    DATE
)
WITH (
    format       = 'PARQUET',
    partitioning = ARRAY['date_key'],
    location     = '<prefix>/sand-box/gravitino/sandbox/gravitino_test_schema/sample_events'
);
```

### Additional context

_No response_

--- a/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
@@ -61,14 +61,7 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
     // Skip import and ownership for staged creates (credential vending).
     // The table will be imported when the staged table is committed via updateTable.
     if (!createTableRequest.stageCreate()) {
-      importTable(context.catalogName(), namespace, createTableRequest.name());
-      IcebergOwnershipUtils.setTableOwner(
-          metalake,
-          context.catalogName(),
-          namespace,
-          createTableRequest.name(),
-          context.userName(),
-          GravitinoEnv.getInstance().ownerDispatcher());
+      importTableAndSetOwner(context, namespace, createTableRequest.name());
     }
 
     return response;
@@ -89,14 +82,7 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
           IcebergIdentifierUtils.toGravitinoTableIdentifier(
               metalake, context.catalogName(), tableIdentifier);
       if (store != null && !store.exists(gravitinoTableId, Entity.EntityType.TABLE)) {
-        importTable(context.catalogName(), tableIdentifier.namespace(), tableIdentifier.name());
-        IcebergOwnershipUtils.setTableOwner(
-            metalake,
-            context.catalogName(),
-            tableIdentifier.namespace(),
-            tableIdentifier.name(),
-            context.userName(),
-            GravitinoEnv.getInstance().ownerDispatcher());
+        importTableAndSetOwner(context, tableIdentifier.namespace(), tableIdentifier.name());
       }
     } catch (IOException ioe) {
       throw new RuntimeException("io exception when checking table entity existence", ioe);
@@ -212,12 +198,20 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
     return dispatcher.getTableMetadataLocation(context, tableIdentifier);
   }
 
-  private void importTable(String catalogName, Namespace namespace, String tableName) {
+  private void importTableAndSetOwner(
+      IcebergRequestContext context, Namespace namespace, String tableName) {
     TableDispatcher tableDispatcher = GravitinoEnv.getInstance().tableDispatcher();
     if (tableDispatcher != null) {
       tableDispatcher.loadTable(
           IcebergIdentifierUtils.toGravitinoTableIdentifier(
-              metalake, catalogName, TableIdentifier.of(namespace, tableName)));
+              metalake, context.catalogName(), TableIdentifier.of(namespace, tableName)));
     }
+    IcebergOwnershipUtils.setTableOwner(
+        metalake,
+        context.catalogName(),
+        namespace,
+        tableName,
+        context.userName(),
+        GravitinoEnv.getInstance().ownerDispatcher());
   }
 }

--- a/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
@@ -58,14 +58,18 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
   public LoadTableResponse createTable(
       IcebergRequestContext context, Namespace namespace, CreateTableRequest createTableRequest) {
     LoadTableResponse response = dispatcher.createTable(context, namespace, createTableRequest);
-    importTable(context.catalogName(), namespace, createTableRequest.name());
-    IcebergOwnershipUtils.setTableOwner(
-        metalake,
-        context.catalogName(),
-        namespace,
-        createTableRequest.name(),
-        context.userName(),
-        GravitinoEnv.getInstance().ownerDispatcher());
+    // Skip import and ownership for staged creates (credential vending).
+    // The table will be imported when the staged table is committed via updateTable.
+    if (!createTableRequest.stageCreate()) {
+      importTable(context.catalogName(), namespace, createTableRequest.name());
+      IcebergOwnershipUtils.setTableOwner(
+          metalake,
+          context.catalogName(),
+          namespace,
+          createTableRequest.name(),
+          context.userName(),
+          GravitinoEnv.getInstance().ownerDispatcher());
+    }
 
     return response;
   }
@@ -75,7 +79,29 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
       IcebergRequestContext context,
       TableIdentifier tableIdentifier,
       UpdateTableRequest updateTableRequest) {
-    return dispatcher.updateTable(context, tableIdentifier, updateTableRequest);
+    LoadTableResponse response =
+        dispatcher.updateTable(context, tableIdentifier, updateTableRequest);
+    // Import the table and set ownership if the entity does not yet exist,
+    // e.g. when committing a staged table create.
+    try {
+      EntityStore store = GravitinoEnv.getInstance().entityStore();
+      NameIdentifier gravitinoTableId =
+          IcebergIdentifierUtils.toGravitinoTableIdentifier(
+              metalake, context.catalogName(), tableIdentifier);
+      if (store != null && !store.exists(gravitinoTableId, Entity.EntityType.TABLE)) {
+        importTable(context.catalogName(), tableIdentifier.namespace(), tableIdentifier.name());
+        IcebergOwnershipUtils.setTableOwner(
+            metalake,
+            context.catalogName(),
+            tableIdentifier.namespace(),
+            tableIdentifier.name(),
+            context.userName(),
+            GravitinoEnv.getInstance().ownerDispatcher());
+      }
+    } catch (IOException ioe) {
+      throw new RuntimeException("io exception when checking table entity existence", ioe);
+    }
+    return response;
   }
 
   @Override

--- a/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
@@ -33,6 +33,7 @@ import org.apache.gravitino.listener.api.event.IcebergRequestContext;
 import org.apache.gravitino.meta.AuditInfo;
 import org.apache.gravitino.meta.TableEntity;
 import org.apache.gravitino.utils.PrincipalUtils;
+import org.apache.iceberg.UpdateRequirement;
 import org.apache.iceberg.catalog.Namespace;
 import org.apache.iceberg.catalog.TableIdentifier;
 import org.apache.iceberg.rest.requests.CreateTableRequest;
@@ -74,18 +75,26 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
       UpdateTableRequest updateTableRequest) {
     LoadTableResponse response =
         dispatcher.updateTable(context, tableIdentifier, updateTableRequest);
-    // Import the table and set ownership if the entity does not yet exist,
-    // e.g. when committing a staged table create.
-    try {
-      EntityStore store = GravitinoEnv.getInstance().entityStore();
-      NameIdentifier gravitinoTableId =
-          IcebergIdentifierUtils.toGravitinoTableIdentifier(
-              metalake, context.catalogName(), tableIdentifier);
-      if (store != null && !store.exists(gravitinoTableId, Entity.EntityType.TABLE)) {
-        importTableAndSetOwner(context, tableIdentifier.namespace(), tableIdentifier.name());
+    // Import the table and set ownership only when committing a staged table create.
+    // A staged create commit is identified by the AssertTableDoesNotExist requirement,
+    // which is set by UpdateRequirements.forCreateTable(). Regular updates (e.g. property
+    // changes) do not carry this requirement, so we must not set ownership for them even
+    // if the entity does not yet exist in the store.
+    boolean isStagedCreateCommit =
+        updateTableRequest.requirements().stream()
+            .anyMatch(UpdateRequirement.AssertTableDoesNotExist.class::isInstance);
+    if (isStagedCreateCommit) {
+      try {
+        EntityStore store = GravitinoEnv.getInstance().entityStore();
+        NameIdentifier gravitinoTableId =
+            IcebergIdentifierUtils.toGravitinoTableIdentifier(
+                metalake, context.catalogName(), tableIdentifier);
+        if (store != null && !store.exists(gravitinoTableId, Entity.EntityType.TABLE)) {
+          importTableAndSetOwner(context, tableIdentifier.namespace(), tableIdentifier.name());
+        }
+      } catch (IOException ioe) {
+        throw new RuntimeException("io exception when checking table entity existence", ioe);
       }
-    } catch (IOException ioe) {
-      throw new RuntimeException("io exception when checking table entity existence", ioe);
     }
     return response;
   }

--- a/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/dispatcher/IcebergTableHookDispatcher.java
@@ -77,24 +77,12 @@ public class IcebergTableHookDispatcher implements IcebergTableOperationDispatch
         dispatcher.updateTable(context, tableIdentifier, updateTableRequest);
     // Import the table and set ownership only when committing a staged table create.
     // A staged create commit is identified by the AssertTableDoesNotExist requirement,
-    // which is set by UpdateRequirements.forCreateTable(). Regular updates (e.g. property
-    // changes) do not carry this requirement, so we must not set ownership for them even
-    // if the entity does not yet exist in the store.
+    // which is set by UpdateRequirements.forCreateTable().
     boolean isStagedCreateCommit =
         updateTableRequest.requirements().stream()
             .anyMatch(UpdateRequirement.AssertTableDoesNotExist.class::isInstance);
     if (isStagedCreateCommit) {
-      try {
-        EntityStore store = GravitinoEnv.getInstance().entityStore();
-        NameIdentifier gravitinoTableId =
-            IcebergIdentifierUtils.toGravitinoTableIdentifier(
-                metalake, context.catalogName(), tableIdentifier);
-        if (store != null && !store.exists(gravitinoTableId, Entity.EntityType.TABLE)) {
-          importTableAndSetOwner(context, tableIdentifier.namespace(), tableIdentifier.name());
-        }
-      } catch (IOException ioe) {
-        throw new RuntimeException("io exception when checking table entity existence", ioe);
-      }
+      importTableAndSetOwner(context, tableIdentifier.namespace(), tableIdentifier.name());
     }
     return response;
   }

--- a/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
@@ -279,7 +279,7 @@ public class TestIcebergTableHookDispatcher {
   }
 
   @Test
-  public void testUpdateTableImportsAndSetsOwnershipForStagedCommit() throws IOException {
+  public void testUpdateTableImportsAndSetsOwnershipForStagedCommit() {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
     // A staged create commit carries AssertTableDoesNotExist as its requirement.
     UpdateTableRequest request =
@@ -289,17 +289,14 @@ public class TestIcebergTableHookDispatcher {
 
     when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
 
-    // Entity does not exist yet (staged table being committed)
-    NameIdentifier gravitinoTableId =
-        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
-    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(false);
-
     LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
 
     Assertions.assertEquals(mockResponse, result);
     verify(mockDispatcher).updateTable(mockContext, tableId, request);
 
     // Verify table import was called
+    NameIdentifier gravitinoTableId =
+        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
     verify(mockTableDispatcher).loadTable(gravitinoTableId);
 
     // Verify ownership was set
@@ -310,66 +307,10 @@ public class TestIcebergTableHookDispatcher {
   }
 
   @Test
-  public void testUpdateTableSkipsImportForStagedCommitWhenEntityAlreadyExists()
-      throws IOException {
-    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    // A staged create commit carries AssertTableDoesNotExist, but the entity already exists
-    // (e.g. idempotent retry). Import and ownership should be skipped.
-    UpdateTableRequest request =
-        new UpdateTableRequest(
-            List.of(new UpdateRequirement.AssertTableDoesNotExist()), Collections.emptyList());
-    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
-
-    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
-
-    NameIdentifier gravitinoTableId =
-        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
-    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(true);
-
-    LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
-
-    Assertions.assertEquals(mockResponse, result);
-    verify(mockDispatcher).updateTable(mockContext, tableId, request);
-
-    // Verify table import was NOT called since entity already exists
-    verify(mockTableDispatcher, never()).loadTable(any());
-
-    // Verify ownership was NOT set since entity already exists
-    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
-  }
-
-  @Test
-  public void testUpdateTableSkipsImportForRegularUpdateWhenEntityMissing() throws IOException {
+  public void testUpdateTableSkipsImportForRegularUpdate() {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
     // Regular table update (e.g. property change): no AssertTableDoesNotExist requirement.
-    UpdateTableRequest request =
-        new UpdateTableRequest(Collections.emptyList(), Collections.emptyList());
-    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
-
-    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
-
-    // Entity does not exist (e.g. table pre-dates Gravitino tracking), but this is NOT a staged
-    // create commit, so we must NOT import or set ownership.
-    NameIdentifier gravitinoTableId =
-        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
-    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(false);
-
-    LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
-
-    Assertions.assertEquals(mockResponse, result);
-    verify(mockDispatcher).updateTable(mockContext, tableId, request);
-
-    // Verify table import was NOT called for a plain property update
-    verify(mockTableDispatcher, never()).loadTable(any());
-
-    // Verify ownership was NOT set
-    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
-  }
-
-  @Test
-  public void testUpdateTablePassesThrough() throws IOException {
-    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    // Regular update without AssertTableDoesNotExist — entity exists.
+    // Import and ownership must NOT be triggered regardless of entity store state.
     UpdateTableRequest request =
         new UpdateTableRequest(Collections.emptyList(), Collections.emptyList());
     LoadTableResponse mockResponse = mock(LoadTableResponse.class);
@@ -381,37 +322,11 @@ public class TestIcebergTableHookDispatcher {
     Assertions.assertEquals(mockResponse, result);
     verify(mockDispatcher).updateTable(mockContext, tableId, request);
 
-    // Verify table import was NOT called
+    // Verify table import was NOT called for a regular update
     verify(mockTableDispatcher, never()).loadTable(any());
 
     // Verify ownership was NOT set
     verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
-  }
-
-  @Test
-  public void testUpdateTableThrowsRuntimeExceptionOnIOException() throws IOException {
-    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    // Use AssertTableDoesNotExist so the code path reaches store.exists().
-    UpdateTableRequest request =
-        new UpdateTableRequest(
-            List.of(new UpdateRequirement.AssertTableDoesNotExist()), Collections.emptyList());
-
-    when(mockDispatcher.updateTable(mockContext, tableId, request))
-        .thenReturn(mock(LoadTableResponse.class));
-
-    NameIdentifier gravitinoTableId =
-        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
-    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE))
-        .thenThrow(new IOException("IO error"));
-
-    RuntimeException exception =
-        Assertions.assertThrows(
-            RuntimeException.class,
-            () -> hookDispatcher.updateTable(mockContext, tableId, request));
-
-    Assertions.assertTrue(
-        exception.getMessage().contains("io exception when checking table entity existence"));
-    verify(mockDispatcher).updateTable(mockContext, tableId, request);
   }
 
   @Test

--- a/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
@@ -30,6 +30,7 @@ import java.io.IOException;
 import java.time.Instant;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import org.apache.commons.lang3.reflect.FieldUtils;
 import org.apache.gravitino.Entity;
 import org.apache.gravitino.EntityStore;
@@ -46,6 +47,7 @@ import org.apache.gravitino.listener.api.event.IcebergRequestContext;
 import org.apache.gravitino.meta.AuditInfo;
 import org.apache.gravitino.meta.TableEntity;
 import org.apache.iceberg.Schema;
+import org.apache.iceberg.UpdateRequirement;
 import org.apache.iceberg.catalog.Namespace;
 import org.apache.iceberg.catalog.TableIdentifier;
 import org.apache.iceberg.rest.requests.CreateTableRequest;
@@ -279,7 +281,10 @@ public class TestIcebergTableHookDispatcher {
   @Test
   public void testUpdateTableImportsAndSetsOwnershipForStagedCommit() throws IOException {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    UpdateTableRequest request = mock(UpdateTableRequest.class);
+    // A staged create commit carries AssertTableDoesNotExist as its requirement.
+    UpdateTableRequest request =
+        new UpdateTableRequest(
+            List.of(new UpdateRequirement.AssertTableDoesNotExist()), Collections.emptyList());
     LoadTableResponse mockResponse = mock(LoadTableResponse.class);
 
     when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
@@ -305,14 +310,18 @@ public class TestIcebergTableHookDispatcher {
   }
 
   @Test
-  public void testUpdateTablePassesThrough() throws IOException {
+  public void testUpdateTableSkipsImportForStagedCommitWhenEntityAlreadyExists()
+      throws IOException {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    UpdateTableRequest request = mock(UpdateTableRequest.class);
+    // A staged create commit carries AssertTableDoesNotExist, but the entity already exists
+    // (e.g. idempotent retry). Import and ownership should be skipped.
+    UpdateTableRequest request =
+        new UpdateTableRequest(
+            List.of(new UpdateRequirement.AssertTableDoesNotExist()), Collections.emptyList());
     LoadTableResponse mockResponse = mock(LoadTableResponse.class);
 
     when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
 
-    // Entity already exists (regular update, not a staged commit)
     NameIdentifier gravitinoTableId =
         IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
     when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(true);
@@ -322,20 +331,73 @@ public class TestIcebergTableHookDispatcher {
     Assertions.assertEquals(mockResponse, result);
     verify(mockDispatcher).updateTable(mockContext, tableId, request);
 
-    // Verify table import was NOT called for existing entity
+    // Verify table import was NOT called since entity already exists
     verify(mockTableDispatcher, never()).loadTable(any());
 
-    // Verify ownership was NOT set for existing entity
+    // Verify ownership was NOT set since entity already exists
+    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
+  }
+
+  @Test
+  public void testUpdateTableSkipsImportForRegularUpdateWhenEntityMissing() throws IOException {
+    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
+    // Regular table update (e.g. property change): no AssertTableDoesNotExist requirement.
+    UpdateTableRequest request =
+        new UpdateTableRequest(Collections.emptyList(), Collections.emptyList());
+    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+
+    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
+
+    // Entity does not exist (e.g. table pre-dates Gravitino tracking), but this is NOT a staged
+    // create commit, so we must NOT import or set ownership.
+    NameIdentifier gravitinoTableId =
+        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
+    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(false);
+
+    LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
+
+    Assertions.assertEquals(mockResponse, result);
+    verify(mockDispatcher).updateTable(mockContext, tableId, request);
+
+    // Verify table import was NOT called for a plain property update
+    verify(mockTableDispatcher, never()).loadTable(any());
+
+    // Verify ownership was NOT set
+    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
+  }
+
+  @Test
+  public void testUpdateTablePassesThrough() throws IOException {
+    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
+    // Regular update without AssertTableDoesNotExist — entity exists.
+    UpdateTableRequest request =
+        new UpdateTableRequest(Collections.emptyList(), Collections.emptyList());
+    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+
+    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
+
+    LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
+
+    Assertions.assertEquals(mockResponse, result);
+    verify(mockDispatcher).updateTable(mockContext, tableId, request);
+
+    // Verify table import was NOT called
+    verify(mockTableDispatcher, never()).loadTable(any());
+
+    // Verify ownership was NOT set
     verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
   }
 
   @Test
   public void testUpdateTableThrowsRuntimeExceptionOnIOException() throws IOException {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
-    UpdateTableRequest request = mock(UpdateTableRequest.class);
-    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+    // Use AssertTableDoesNotExist so the code path reaches store.exists().
+    UpdateTableRequest request =
+        new UpdateTableRequest(
+            List.of(new UpdateRequirement.AssertTableDoesNotExist()), Collections.emptyList());
 
-    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
+    when(mockDispatcher.updateTable(mockContext, tableId, request))
+        .thenReturn(mock(LoadTableResponse.class));
 
     NameIdentifier gravitinoTableId =
         IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);

--- a/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
+++ b/iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/dispatcher/TestIcebergTableHookDispatcher.java
@@ -22,6 +22,7 @@ import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
@@ -251,16 +252,103 @@ public class TestIcebergTableHookDispatcher {
   }
 
   @Test
-  public void testUpdateTablePassesThrough() {
+  public void testCreateTableSkipsImportAndOwnershipForStageCreate() {
+    Namespace namespace = Namespace.of("test_schema");
+    CreateTableRequest request =
+        CreateTableRequest.builder()
+            .withName("test_table")
+            .withSchema(TABLE_SCHEMA)
+            .stageCreate()
+            .build();
+
+    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+    when(mockDispatcher.createTable(mockContext, namespace, request)).thenReturn(mockResponse);
+
+    LoadTableResponse result = hookDispatcher.createTable(mockContext, namespace, request);
+
+    Assertions.assertEquals(mockResponse, result);
+    verify(mockDispatcher).createTable(mockContext, namespace, request);
+
+    // Verify table import was NOT called for staged create
+    verify(mockTableDispatcher, never()).loadTable(any());
+
+    // Verify ownership was NOT set for staged create
+    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
+  }
+
+  @Test
+  public void testUpdateTableImportsAndSetsOwnershipForStagedCommit() throws IOException {
     TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
     UpdateTableRequest request = mock(UpdateTableRequest.class);
     LoadTableResponse mockResponse = mock(LoadTableResponse.class);
 
     when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
 
+    // Entity does not exist yet (staged table being committed)
+    NameIdentifier gravitinoTableId =
+        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
+    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(false);
+
     LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
 
     Assertions.assertEquals(mockResponse, result);
+    verify(mockDispatcher).updateTable(mockContext, tableId, request);
+
+    // Verify table import was called
+    verify(mockTableDispatcher).loadTable(gravitinoTableId);
+
+    // Verify ownership was set
+    ArgumentCaptor<String> userCaptor = ArgumentCaptor.forClass(String.class);
+    verify(mockOwnerDispatcher)
+        .setOwner(eq(TEST_METALAKE), any(), userCaptor.capture(), eq(Owner.Type.USER));
+    Assertions.assertEquals(TEST_USER, userCaptor.getValue());
+  }
+
+  @Test
+  public void testUpdateTablePassesThrough() throws IOException {
+    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
+    UpdateTableRequest request = mock(UpdateTableRequest.class);
+    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+
+    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
+
+    // Entity already exists (regular update, not a staged commit)
+    NameIdentifier gravitinoTableId =
+        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
+    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE)).thenReturn(true);
+
+    LoadTableResponse result = hookDispatcher.updateTable(mockContext, tableId, request);
+
+    Assertions.assertEquals(mockResponse, result);
+    verify(mockDispatcher).updateTable(mockContext, tableId, request);
+
+    // Verify table import was NOT called for existing entity
+    verify(mockTableDispatcher, never()).loadTable(any());
+
+    // Verify ownership was NOT set for existing entity
+    verify(mockOwnerDispatcher, never()).setOwner(any(), any(), any(), any());
+  }
+
+  @Test
+  public void testUpdateTableThrowsRuntimeExceptionOnIOException() throws IOException {
+    TableIdentifier tableId = TableIdentifier.of("test_schema", "test_table");
+    UpdateTableRequest request = mock(UpdateTableRequest.class);
+    LoadTableResponse mockResponse = mock(LoadTableResponse.class);
+
+    when(mockDispatcher.updateTable(mockContext, tableId, request)).thenReturn(mockResponse);
+
+    NameIdentifier gravitinoTableId =
+        IcebergIdentifierUtils.toGravitinoTableIdentifier(TEST_METALAKE, TEST_CATALOG, tableId);
+    when(mockEntityStore.exists(gravitinoTableId, Entity.EntityType.TABLE))
+        .thenThrow(new IOException("IO error"));
+
+    RuntimeException exception =
+        Assertions.assertThrows(
+            RuntimeException.class,
+            () -> hookDispatcher.updateTable(mockContext, tableId, request));
+
+    Assertions.assertTrue(
+        exception.getMessage().contains("io exception when checking table entity existence"));
     verify(mockDispatcher).updateTable(mockContext, tableId, request);
   }
 

